hot-directory/src/caching_directory.rs

```diff
@@ -0,0 +1,336 @@
+/*
+    Quickwit
+    Copyright (C) 2021 Quickwit Inc.
+
+    Quickwit is offered under the AGPL v3.0 and as commercial software.
+    For commercial licensing, contact us at hello@quickwit.io.
+
+    AGPL:
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use async_trait::async_trait;
+use lru::LruCache;
+use std::fmt;
+use std::io;
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Mutex};
+use tantivy::directory::error::{DeleteError, OpenReadError, OpenWriteError};
+use tantivy::directory::{FileHandle, OwnedBytes, WatchHandle, WritePtr};
+use tantivy::{AsyncIoResult, Directory, HasLen};
+use tracing::{info, warn};
+
+#[derive(Hash, Debug, Clone, PartialEq, Eq)]
+struct SliceAddress {
+    path: PathBuf,
```

Comment:
General comment about `hot-directory`.

We can merge it now, but I suspect it does not meet code quality. I'll do a pass over it.